### PR TITLE
conform valuesFile & valuesFiles implementations and add test cases

### PR DESCRIPTION
**Pull request by nniikkoollaaii**
Hey @arbourd 

I'm hoping here are some improvements to your valuesFiles implementation at PR https://github.com/fluxcd/source-controller/pull/305.

I found the following problems:

valuesFile/vaulesFiles are used in two functions: [reconcileFromHelmRepository()](https://github.com/arbourd/source-controller/blob/main/controllers/helmchart_controller.go#L296) and [reconcileFromTarballArtifact()](https://github.com/arbourd/source-controller/blob/main/controllers/helmchart_controller.go#L489)

For reconcileFromTarballArtifact(): 

- The untar operation at Line [505](https://github.com/arbourd/source-controller/blob/main/controllers/helmchart_controller.go#L505) does NOT untar the helmchart to the tmpDir according to my tests. There is only a helmchart-0.1.0.tgz file afterwards and the Line [522](https://github.com/arbourd/source-controller/blob/main/controllers/helmchart_controller.go#L522) is reading the helmchart from this .tgz-File.
So your implementation reading the valuesFile(s) from the tmpDir is not working as there is no `helmchart/override.yaml` in the tmpDir.
I suspect the original implementation wasn't functional too.
I was able to test your implementation after I added an additional untar step around Line 553. Then your implementation successful read the valuesFiles files.

For reconcileFromHelmRepository():

- There were no testcases for this usage of valuesFile in the original code. I added some [here](https://github.com/nniikkoollaaii/source-controller/blob/values-files/controllers/helmchart_controller_test.go#L150)

- It seems like you copy pasted your valuesFiles implementation from reconcileFromTarballArtifact(). (I know this was your WIP). But this cannot work because in reconcileFromHelmRepository() the helmchart gets never unpacked to a tmpDir. The original [implementation](https://github.com/fluxcd/source-controller/blob/main/controllers/helmchart_controller.go#L389) read the helmchart from a `bytes.Buffer` and accessed the values files from `helmchart.Files`.
So for this implementation you have to specify the path "./override.yaml".
For your implementation at reconcileFromTarballArtifact() you have to specify the path "./helmchart/override.yaml" relativ to the tmpDir root.
This does not look nice to me.

In addition I updated the testcases to check for the generated values like [here](https://github.com/nniikkoollaaii/source-controller/blob/values-files/controllers/helmchart_controller_test.go#L192) as suggested in the original PR

So I implemented this functionality at both places via `helmchart.Files` at Line [402](https://github.com/nniikkoollaaii/source-controller/blob/values-files/controllers/helmchart_controller.go#L402) and [554](https://github.com/nniikkoollaaii/source-controller/blob/values-files/controllers/helmchart_controller.go#L554).

I have one problem left: All three test cases `When("Setting invalid valuesFile attribute", func() { ...` are failing with `Message: "Operation cannot be fulfilled on helmcharts.source.toolkit.fluxcd.io \"helmchart-sample-81w8c\": the object has been modified; please apply your changes to the latest version and try again",`.

I currently don't understand what is causing this :/. I'm hoping you could help here?

### CODE REVIEW

**Comment by arbourd** [comment]
Wow this is awesome! Thank you for helping me out with this. Will take a look this weekend 😁

**Comment by arbourd** [comment]
> The untar operation at Line 505 does NOT untar the helmchart to the tmpDir according to my tests. There is only a helmchart-0.1.0.tgz file afterwards and the Line 522 is reading the helmchart from this .tgz-File.

Would like to know how you got that impression, the tests? We're actually running this PR in a test environment and I can confirm that it works (using `ValuesFiles`).

> For your implementation at reconcileFromTarballArtifact() you have to specify the path "./helmchart/override.yaml" relativ to the tmpDir root.

I agree this doesn't look nice.

File: controllers/helmchart_controller.go [hunk 1]
```diff
@@ -505,22 +541,53 @@ func (r *HelmChartReconciler) reconcileFromTarballArtifact(ctx context.Context,
...
+
+			// Find override file and retrieve contents
+			var valuesData []byte
+			cfn := filepath.Clean(v)
+			for _, f := range helmChart.Files {
+				if f.Name == cfn {
+					valuesData = f.Data
+					break
+				}
+			}
```
Comment (arbourd) [inline controllers/helmchart_controller.go:564]: I do like the idea of loading values data from the already-in-memory `helmChart`

**Comment by nniikkoollaaii** [comment]
> > The untar operation at Line 505 does NOT untar the helmchart to the tmpDir according to my tests. There is only a helmchart-0.1.0.tgz file afterwards and the Line 522 is reading the helmchart from this .tgz-File.
> 
> Would like to know how you got that impression, the tests? We're actually running this PR in a test environment and I can confirm that it works (using `ValuesFiles`).
> 
> > For your implementation at reconcileFromTarballArtifact() you have to specify the path "./helmchart/override.yaml" relativ to the tmpDir root.
> 
> I agree this doesn't look nice.

Then I've made a mistake during my tests. 
I commented out the `defer os.RemoveAll(tmpDir)` at line [495](https://github.com/arbourd/source-controller/blob/values-files/controllers/helmchart_controller.go#L495) and printed out the `tmpDir` value during test runtime and looked at it afterwards on my filesystem. 
In addition I added a `fmt.Println(chartPath)` after Line [511](https://github.com/arbourd/source-controller/blob/values-files/controllers/helmchart_controller.go#L511)

But this is probably due to the unit test setup.

**Comment by arbourd** [comment]
I'm going to close this because we've solved these issues in https://github.com/fluxcd/source-controller/pull/332.

Thanks a lot for your help @nniikkoollaaii. Couldn't have done it without you 🎉

**Pull request closed without merging**